Added encoding to user name and password

taylorstine opened the pull request:
I have an @ symbol in my mongo password. I realize that I can specify a custom class for the DatabaseAdapter, but I believe the decoding of a username and password should be built in to the default implementation.

gfosco commented on ExportAdapter.js:
```diff
@@ -34,8 +34,19 @@ ExportAdapter.prototype.connect = function() {
+  var lastAtIndex = this.mongoURI.lastIndexOf('@');
+  var encodedMongoURI = this.mongoURI;
+  var username = null;
+  var password = null;
```
It does not look like `username` or `password` are used... and it looks like the database uri is being somewhat validated here, so how about we throw an error if it's not?

taylorstine commented:
@gfosco sorry for that, I've removed old references.  I've added a regex validation for the mongo URI.  It won't capture all invalid URIs, but it will at least validate the monogURI to be in the form we expect, [as specified here](https://docs.mongodb.org/manual/reference/connection-string/).

gfosco commented:
Thanks for adding the regxr link.. very cool. :+1:

themichael123 commented:
This causes MongoDB Replica Sets to no longer work.

taylorstine commented:
@themichael123 can you provide an example URI?

themichael123 commented:
@taylorstine Have a look here: https://docs.mongodb.org/manual/reference/connection-string/

I wouldn't use regex for this task.

adamontherun commented:
@taylorstine @gfosco here's an example of a paid MongoLabs URI that isn't working: mongodb://test:testpass@ds056315-a0.mongolab.com:59325,ds059315-a1.mongolab.com:59315/testDBname?replicaSet=rs-ds059415

here's an example of a free sandbox MongoLabs URI that is working:
mongodb://test:testpass@ds053145.mongolab.com:59145/mytestserver

adamontherun commented:
I confirmed that the MongoLabs replica set URI worked fine in Parse Server 2.0.4, the breaking change was in 2.0.5

gfosco commented:
Let's get a fix PR sent in to parse-server to address this... @taylorstine are you interested in doing it?

taylorstine commented:
@gfosco sure, but I could use a second opinion on the regex.  My aptitude for regex has kind of reached it's limit, maybe it should just be removed?

adamontherun commented:
@gfosco @taylorstine I'd vote for removing the regex until we can be very confident in it. this bug breaks most parse server deployments that aren't running in the sandbox.

drew-gross commented:
The approach we used on hosted parse was to remove one character at a time from the end of the connection string until it looked like a regular URI, and then we could use a regular URI parsing library on it, modify the contents as needed, and then rebuild the connection string that includes all the replica sets. That approach might work here as well. Make sure to add tests if you use that approach though.

flovilmart commented:
We could also split the host of the URI around "@" char, and leave the last element being the host, and reconcat the n-1 first elements, then split on ":".

jpgupta commented:
In my case it's moving from a sandbox to a production environment on MongoLabs that breaks it. 

It appears to be that the additional ?replicaSet argument added onto the URI fails the regex expression, I'd suggest taking it out as it's a little over-engineered. Testing the Mongo connection success/failure is probably worth while but it seems better to just trust people that they've entered a valid uri and will preserve future changes to the URI structure / arguments

taylorstine commented:
https://github.com/ParsePlatform/parse-server/pull/297 I'm not confident in this approach, I don't want to break anything just because my password has a `@` sign

flovilmart commented:
So could we split the full host on "@", then take the last part as the 'real' host and join the [0, n-1] parts together? Then from that, we split on ":" to get the username / password.
